test/unit/ngsiv2/general/deviceService-test.js

```diff
@@ -0,0 +1,261 @@
+                    }
+                ],
+                active: [
+                    {
+                        name: 'humidity',
+                        type: 'percentage'
+                    }
+                ]
+            }
+        },
+        iotManager: {
+            host: 'localhost',
+            port: 8082,
+            path: '/protocols',
+            protocol: 'MQTT_UL',
+            description: 'MQTT Ultralight 2.0 IoT Agent (Node.js version)'
+        },
+        service: 'smartGondor',
+        subservice: 'gardens',
+        providerUrl: 'http://smartGondor.com',
+        deviceRegistrationDuration: 'P1M',
+        throttling: 'PT5S'
+    },
+    groupCreation = {
+        url: 'http://localhost:' + iotAgentConfig.server.port + '/iot/services',
+        method: 'POST',
+        json: {
+            services: [
+                {
+                    resource: '',
+                    apikey: '801230BJKL23Y9090DSFL123HJK09H324HV8732',
+                    entity_type: 'TheLightType',
+                    trust: '8970A9078A803H3BL98PINEQRW8342HBAMS',
+                    cbHost: 'http://unexistentHost:1026',
+                    commands: [],
+                    lazy: [],
+                    attributes: [
+                        {
+                            name: 'status',
+                            type: 'Boolean'
+                        }
+                    ],
+                    static_attributes: []
+                }
+            ]
+        },
+        headers: {
+            'fiware-service': 'TestService',
+            'fiware-servicepath': '/testingPath'
+        }
+    },
+    deviceCreation = {
+        url: 'http://localhost:' + iotAgentConfig.server.port + '/iot/devices',
+        method: 'POST',
+        json: utils.readExampleFile('./test/unit/examples/deviceProvisioningRequests/provisionNewDevice.json'),
+        headers: {
+            'fiware-service': 'TestService',
+            'fiware-servicepath': '/testingPath'
+        }
+    },
+    contextBrokerMock,
+    iotamMock;
+
+
+/* jshint camelcase: false */
+describe('Device Service: utils', function() {
+    beforeEach(function(done) {
+        nock.cleanAll();
+        logger.setLevel('FATAL');
+        iotamMock = nock('http://localhost:8082')
+            .post('/protocols')
+            .reply(200, {});
+
+        iotAgentLib.activate(iotAgentConfig, done);
+    });
+
+    afterEach(function(done) {
+        nock.cleanAll();
+        async.series([
+            iotAgentLib.clearAll,
+            iotAgentLib.deactivate
+        ], done);
+    });
+
+    describe('When an existing device tries to be retrieved with retrieveOrCreate()', function() {
+        beforeEach(function(done) {
+            contextBrokerMock = nock('http://192.168.1.1:1026')
+                .matchHeader('fiware-service', 'TestService')
+                .matchHeader('fiware-servicepath', '/testingPath')
+                .post('/v2/entities?options=upsert')
```

Comment:
This is quite common in a great number of unit tests of the IOTAs. It is true that this type of mock does not allow checking if the payload sent to the Orion CB is correct but I understand that the objective of this test is not to verify the device provisioning functionality but, as the describe/it sentences state `When an existing device tries to be retrieved with retrieveOrCreate() should return the existing device`. The test is focused on verifying the `retrieveDevice` function and therefore for simplicity does not include the specification of the complete mock for the device provisioning.

Reply:
It get your point. However if you are not "in deep" in this code (as me :) it may seem to be a bit weird. A comment in this case may help but, if you think it wouldn't make sense to to that and that the describe() messages are enough, then just mark this comment thread with NTC.

Reply:
I have included a comment in 0720dc5339cd2b1e0487c37f2e48a76f057f4101. What do you think?

Reply:
Comment looks nice :) 

My only doubt if this this situation happens in other points in the code and more commets like that are needed... I'll review the mocks in the PR and provide additional comments. With regards to this particular one, NTC after that commit.